File: OOP/OOP/source_files/group.cpp
#include <iostream>
#include <fstream>
#include "../header_files/group.h"

using namespace std;


Group::Group(const std::string &title){
    this->title = title;
}


std::string Group::get_title() const{
    return this->title;
}


void Group::showAllStudents() const{
    if (!this->students.size()){
        cout << "Нет обучающихся в группе!" << endl;
        return;
    }
        
    for (const auto& student : this->students){
        cout << "-----" << student->get_className() << " " << student->get_id() << "-----" << endl;
        student->writeToConsole();
        cout << "--------------" << endl;
    }
        
}


void Group::deleteAllStudents(){
    this->students.clear();
    cout << "все обучающиеся удалены!" << endl;
}


void Group::saveStudents(std::ostream &file){
    boost::archive::binary_oarchive ofile(file);
    ofile << *this;
    
    cout << "все обучающиеся сохранены!" << endl;
}


void Group::loadStudents(){
    string file_name;
    
    cout << "файл для загрузки: ";
    getline(cin>>std::ws, file_name);
    
    ifstream file("./static/" + file_name);
    
    if (file.is_open()){
        boost::archive::binary_iarchive ifile(file);
        ifile >> *this;
        this->set_current_studentID();
    } else {
        cout << "файл не открыт!" << endl;
    }
    
    cout << this->students.size() << " обучающихся было загружено" << endl;
}


void Group::set_current_studentID(){
    int max_id = 0;
    
    for (const auto &student : this->students){
        max_id = max(max_id, student->get_id());
    }
    
    Student::set_currentID(max_id);
}


Group::~Group(){
    this->deleteAllStudents();
}

File: SHARP/StudentsManager/Group.cs
using System;
using System.Text.Json;
using System.Text.Json.Serialization;


namespace StudentManager
{
    public class Group
    {
        public string Title {get; set;}
        public List<Student> Students {get; set;} = new List<Student>();

        public Group(string title)
        {
            Title = title;
        }

        public void AddStudent(Student newStudent)
        {
            Students.Add(newStudent);
        }

        public void ShowAllStudents()
        {
            if (Students.Count == 0)
            {
                Console.WriteLine("нет студентов");
                return;
            }

            foreach(var student in Students)
            {
                Console.WriteLine($"-----{student.ClassName} {student.StudentId}----");
                student.WriteToConsole();
                Console.WriteLine("--------------");
            }
        }

        public void DeleteAllStudents()
        {
            Students.Clear();
            Console.WriteLine("все обучающиеся удалены");
        }

        private string _getPath()
        {
            Console.Write("имя файла: ");

            string? input = Console.ReadLine();
            string fileName = !string.IsNullOrEmpty(input) ? input : "data.json";

            string directory = "static";
            if (!Directory.Exists(directory))
            {
                Directory.CreateDirectory(directory);
            }

            string filePath = Path.Combine(directory, fileName);

            return filePath;
        }

        public void SaveAllStudents()
        {
            try
            {
                string filePath = _getPath();
                
                var options = new JsonSerializerOptions
                {
                    WriteIndented = true,
                };

                string json = JsonSerializer.Serialize(Students, options);
                File.WriteAllText(filePath, json);
                
                Console.WriteLine("Все обучающиеся сохранены!");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Не удалось сохранить файл: {ex.Message}");
            }
        }

        public void LoadStudents()
        {
            try
            {
                string filePath = _getPath();
                
                if (!File.Exists(filePath))
                {
                    Console.WriteLine("Файл не существует!");
                    return;
                }

                string json = File.ReadAllText(filePath);
                
                var loadedStudents = JsonSerializer.Deserialize<List<Student>>(json);
                
                if (loadedStudents != null)
                {
                    Students = loadedStudents;
                    
                    int maxId = Students.Count > 0 ? Students.Max(s => s.StudentId) : 0;
                    Student.SetCurrentId(maxId);
                    
                    Console.WriteLine($"{Students.Count} обучающихся было загружено");
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Не удалось загрузить файл: {ex.Message}");
            }
        }
    }
}
